apps/passport-server/src/apis/lemonade/lemonadeAPI.ts

```diff
@@ -1,78 +1,241 @@
+   * previously-fetched results.
+   */
+  private async paginate<T>(
+    query: ({ skip, limit }: { skip: number; limit: number }) => Promise<T[]>,
+    limit = 25
+  ): Promise<T[]> {
+    const results = [];
+    let page = [];
+
+    // Lemonade's pagination approach works by setting an offset ('skip') into
+    // the collection being fetched. This is vulnerable to the collection
+    // changing between requests, but there's not much we can do about that.
+    do {
+      page = await query({
+        // Skip over previously-fetched events
+        skip: results.length,
+        // Try to fetch `limit` number of events
+        limit
+      });
+      results.push(...page);
+      // If we received a full page, keep going, otherwise stop here
+    } while (page.length === limit);
+
+    return results;
+  }
+
+  /**
+   * Get all events (restricted by credentials).
+   */
+  public async getHostingEvents(
+    backendUrl: string,
+    credentials: LemonadeOAuthCredentials
+  ): Promise<LemonadeEvents> {
+    const client = this.getClient(backendUrl);
+    const token = await this.getToken(credentials);
+
+    return await this.paginate<LemonadeEvents[number]>((opts) =>
+      client.getHostingEvents(token, opts)
+    );
+  }
+
+  /**
+   * Get all ticket types for an event.
+   */
+  public async getEventTicketTypes(
+    backendUrl: string,
+    credentials: LemonadeOAuthCredentials,
+    lemonadeEventId: string
+  ): Promise<LemonadeTicketTypes> {
+    const client = this.getClient(backendUrl);
+    const token = await this.getToken(credentials);
+    return await client.getEventTicketTypes(token, {
+      input: { event: lemonadeEventId }
+    });
+  }
+
+  /**
+   * Get all tickets for an event.
+   * Unlike the other calls, this does not use the GraphQL client.
+   * Instead, it reads CSV data from a REST endpoint.
+   */
+  public async getTickets(
+    backendUrl: string,
+    credentials: LemonadeOAuthCredentials,
+    lemonadeEventId: string
+  ): Promise<LemonadeTicket[]> {
+    let token = await this.getToken(credentials);
+    const url = urljoin(backendUrl, "event", lemonadeEventId, "export/tickets");
+    let result = await instrumentedFetch(url, {
+      method: "POST",
+      headers: {
+        Authorization: `Bearer ${token}`,
+        "Content-type": "text/csv"
+      }
+    });
+    if (result.status === 401) {
+      // If we received an authentication failure, try again once with a new
+      // token
+      this.invalidateToken(credentials);
+      token = await this.getToken(credentials);
+      result = await instrumentedFetch(url, {
+        method: "POST",
+        headers: {
+          Authorization: `Bearer ${token}`,
+          "Content-type": "text/csv"
+        }
+      });
+    }
+    const csvText = await result.text();
+    const parsed = parse(csvText, {
+      columns: true
+    });
+    const results = [];
+    for (const row of parsed) {
+      results.push(LemonadeTicketSchema.parse(row));
```

Comment:
I've moved parsing to the Pipeline level. This means that parsing failures are limited to individual tickets.

In practice, it's pretty hard for a Lemonade ticket to fail to parse, because some fields are genuinely optional. For instance, we get tickets which have no user email or user ID, because these are tickets for people who are invited to the event but have not yet registered with Lemonade. We skip these during load because they can't be checked in. I've added a test to demonstrate this.

However, ticket type is mandatory and I'm not aware of a case where ticket type can legitimately be empty. I've also added a test for a case where a ticket has no ticket type, to show that it is ignored.